### PR TITLE
[#8] User Setting Domain Model

### PR DESCRIPTION
**Pull request by sueelly**
## #️⃣ Related Issue

Linked Feature backlog: #8 
## 📝 Purpose of PR

유저 도메인의 설정 모델을 정의한다.

## Contents

도메인 모델은 생성자가 아닌 builder 사용을 강제한다.
도메인 모델에서는 모델의 각 필드에 대한 유효성 검증이 진행된다.
도메인 모델은 외부 의존성을 갖지 않으며 (도메인 모델의 완전성), 도메인 서비스의 경우 spring boot의 @Service가 아닌 여러 도메인을 종합하여 처리하는 도메인 로직을 구현하는 데에만 사용된다.
결과적으로 도메인 모듈은 어떤 의존성도 갖지 않게 된다.

### 고려사항 1. 도메인 모델의 외부 의존성
Domain Driven Development에서는 도메인 모델에서 모든 도메인 로직을 처리해야한다고 말한다 (도메인 순수성).
헌데 그 경우에 외부 의존성이 필요할 때 (대표적으로 DB read / write), 어떻게 처리할지 궁금했다.
결과적으로 모든 결론은 동일했다. 어플리케이션 서비스에서 처리하는 것.
외부 의존성을 가져도 되는 곳에서 처리하는 것이다.
다만 도메인 로직, 즉 도메인의 어떤 값에 대한 '결정'을 내리는 것을 도메인 모델에 위임하는 것이다.

https://www.borntodare.me/ef699637-514c-469e-9a2f-32718b911b72
https://www.borntodare.me/df4498f8-b82c-4a64-9be6-8acbcceef3f2
위의 2가지 글을 참고하였다.

### 고려사항 2. DDD...?
기본적인 CRUD는 도메인 로직이랄 게 따로 없는 빈약한 도메인 모델만으로도 충분히 구현가능하다고 한다. 그리고 SnapNOW는 빈약한 도메인 모델로도 충분한 것 같다는 생각이 들었다.
DDD에 발을 담구게된 계기는 다음과 같았다.
1. 각 역할을 분명히 하고, 의존성에 대해 상세히 관리하기 위해 api, domain, infra 모듈을 분리했다.
2. 모듈별 데이터 모델을 따로 정의하였으며, 이들을 있는 converter class를 두었다.
3. 각 모듈, 각 계층이 확실하게 분리된 상황에서 헥사고날을 접하였고, 도메인을 가장 중심에 두어 의존성을 갖지 않도록 설계하는 것이 유지보수에 가장 좋을 것 같다는 생각을 했다. application 모듈이 바뀌었을 때, infra가 바뀌었을 때도 도메인은 그대로 유지되어야한다는 생각이 들었기 때문이다. 그렇게 useCase와 repository를 사용하기로 했다.
4. useCase로 다 정의되지 못하는 서비스만의 메서드가 존재하는 경우, 해당 메서드를 다른 곳에서 호출할 시에 의존성이 인터페이스를 갖는 게 아니라 서비스를 직접적으로 갖게 된다. 이에 대해 어떻게할지 고민하던 차에 '도메인 모델로 모든 도메인 로직을 처리'한다는 DDD의 개념을 접하고 이런 식으로 구현하면 문제가 해결되지 않을까했다.
5. 또한 헥사고날은 DDD와 함께 많이 쓰이는 형태였고, domain 중심으로만 현재 코드도 구현되고 있는 상황이기에 적절하다고 생각했었다. 또한 도메인별로 잘 분리해두면 추후 MSA로 변환하기도 좋을 것이라 생각했다.

하지만 user, post (추후 notification, chatting 추가 예정)이 주된 도메인인 SnapNOW의 경우에 '도메인 중심'이 맞는지에 대한 생각이 들었다.
도메인 모델이 풍부하지 않기에, application에서 도메인 모델을 그렇게 활용을 잘 하나라는 생각도 들었다...

다만 그럼에도 적용해보고 싶은 것은 여러 서버로 쪼개기 좋다는 점과, Post 도메인에 대해 CQRS 패턴을 적용해보고 싶기 때문에 (실은 push / pull model도 이에 포함된다고 생각했다) 우선 계속 유지하여 개발을 진행할 예정이다.

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: snapnow-domain/src/main/java/kr/flab/snapnow/domain/user/domain/enums/setting/SettingCategory.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        EnumSet.allOf(PrivacySettingType.class),
+        EnumSet.allOf(PrivacySettingOption.class),
+        PrivacySettingOption.PUBLIC);
+
+    private final String value;
+    private final EnumSet<? extends SettingType> settingType;
+    private final EnumSet<? extends SettingOption> settingOption;
+    private final SettingOption defaultSettingOption;
+
+    public static SettingCategory of(String value) {
```
Comment (f-lab-jason) [inline snapnow-domain/src/main/java/kr/flab/snapnow/domain/user/domain/enums/setting/SettingCategory.java:36]: enum type인데 `of`가 필요한건 어떤 케이스인가요?

**Review by f-lab-jason — COMMENTED** [review]
`SettingCategory`가 굳이 필요한 이유는 무엇인가요?
별개의 설정이면 굳이 묶어서 복잡도를 증가시킬 필요가 없어보여서요.
Type과 Option을 분리할 필요도 잘 와닿지 않는데 설명해주시면 좋을 것 같아요!

File: snapnow-domain/src/main/java/kr/flab/snapnow/domain/user/domain/enums/setting/SettingCategory.java (hunk 1 above)
Comment (sueelly) [reply snapnow-domain/src/main/java/kr/flab/snapnow/domain/user/domain/enums/setting/SettingCategory.java:36]: **SettingCategory 생성 이유**
1. 유한한 값의 집합을 나타낼 클래스는 enum으로 선언하였습니다. 따라서 설정 카테고리 또한 유한한 카테고리들의 집합이므로  enum이 있어야한다고 판단했습니다. 설정들 리스트 또한 서버에서 보내줘야한다는 의미에서도 필요하다고 생각했습니다.
2. 현재는 계정 공개 설정과 알림 설정만 존재하는데, 추후 카테고리가 많이 추가될 수 있을 거라 생각했습니다. 값을 추가함에 있어서도 enum으로 선언하는 게 편리하다고 생각했습니다.
3. 카테고리별 설정할 수 있는 타입과 선택할 수 있는 옵션이 다 다르기 때문에, 카테고리에 연관된 정보도 한번에 다루기 위해 필요하다고 생각했습니다.

**Type과 option을 분리한 이유**
우선 전제는 각 카테고리별 다른 타입과 옵션을 갖는 상황에서, 한 카테고리의 서로 다른 타입들은 같은 설정 옵션을 갖는다 입니다. 
알림 설정의 경우 comment, like, mention, new follow라는 타입이 있는데, 모두 all, follow, off라는 옵션을 갖습니다.
분리하지 않는 경우에는 각 타입별 옵션값을 일일이 설정해줘야하는데, 동일하게 반복된다면 따로 정의하는 게 나을 거라고 생각했습니다

**of가 필요한 케이스**
클라이언트로부터 요청을 받을 때, value 형태로 값을 받아온다고 생각했습니다.
클라이언트에 카테고리와 타입, 옵션에 대한 데이터를 보낼 때 화면에 표시할 value만을 보내고, 그 값으로 요청을 받을 테니 value로부터 enum을 반환할 수 있는 of가 필요할 것이라 생각했습니다.
다만 그렇게 한다고 했을 때 현재 코드 상에서 type과 option에 대해서도 동일하게 적용했어야했을 것 같네요.
그대로 진행하라 하시면 추가하도록 하겠습니다.
